src/Features/LanguageServer/Protocol/Handler/Completion/CompletionHandler.cs

```diff
@@ -401,6 +415,212 @@ static void PromoteCommonCommitCharactersOntoList(LSP.VSCompletionList completio
+            var index = 0;
+            var completionHelper = CompletionHelper.GetHelper(document);
+            foreach (var item in completionList.Items)
+            {
+                if (CompletionHelper.TryCreateMatchResult<CompletionItem?>(
+                    completionHelper,
+                    item,
+                    editorCompletionItem: null,
+                    filterText,
+                    completionTrigger.Kind,
+                    GetFilterReason(completionTrigger),
+                    recentItems: ImmutableArray<string>.Empty,
+                    includeMatchSpans: false,
+                    index,
+                    out var matchResult))
+                {
+                    matchResultsBuilder.Add(matchResult);
+                    index++;
+                }
+            }
+
+            // Next, we sort the list based on the pattern matching result.
+            matchResultsBuilder.Sort(MatchResult<CompletionItem?>.SortingComparer);
+
+            // Finally, truncate the list to 1000 items plus any preselected items that occur after the first 1000.
+            var filteredList = matchResultsBuilder
+                .Take(MaxCompletionListSize)
+                .Concat(matchResultsBuilder.Skip(MaxCompletionListSize).Where(match => ShouldItemBePreselected(match.RoslynCompletionItem)))
+                .Select(matchResult => matchResult.RoslynCompletionItem)
+                .ToImmutableArray();
+            var newCompletionList = completionList.WithItems(filteredList);
+
+            // Per the LSP spec, the completion list should be marked with isIncomplete = false when further insertions will
+            // not generate any more completion items.  This means that we should be checking if the matchedResults is larger
+            // than the filteredList.  However, the VS client has a bug where they do not properly re-trigger completion
+            // when a character is deleted to go from a complete list back to an incomplete list.
+            // For example, the following scenario.
+            // User types "So" -> server gives subset of items for "So" with isIncomplete = true
+            // User types "m" -> server gives entire set of items for "Som" with isIncomplete = false
+            // User deletes "m" -> client has to remember that "So" results were incomplete and re-request if the user types something else, like "n"
+            //
+            // Currently the VS client does not remember to re-request, so the completion list only ever shows items from "Som"
+            // so we always set the isIncomplete flag to true when the original list size (computed when no filter text was typed) is too large.
+            // VS bug here - https://devdiv.visualstudio.com/DevDiv/_workitems/edit/1335142
+            var isIncomplete = completionList.Items.Length > newCompletionList.Items.Length;
+
+            return (newCompletionList, isIncomplete);
```

Comment:
i'm a little surprised i don't see the remainder of the results cached anywhere.  how are they determined when the client calls back in?

Reply:
discussed offline, recomputing until theres LSP support for passing result ids between server and client